tests/test_util/test_warning.py

```diff
@@ -1,38 +1,88 @@
 #!/usr/bin/env python
 
-"""Unit tests for union_dict.
-"""
-from unittest import TestCase, main
+import pickle
 
-from cogent3.util.warning import deprecate
+import pytest
 
+from cogent3.util.warning import deprecated_args
 
-__author__ = "Richard Morris"
-__copyright__ = "Copyright 2007-2022, The Cogent Project"
-__credits__ = ["Gavin Huttley", "Richard Morris"]
-__license__ = "BSD-3"
-__version__ = "2023.2.12a1"
-__maintainer__ = "Gavin Huttley"
-__email__ = "gavin.huttley@anu.edu.au"
-__status__ = "Production"
 
+def test_deprecated_args():
+    # Example target function to be decorated
+    @deprecated_args(
+        [("x", "a"), ("y", "b")],
+        version="a future release",
+        reason="x and y are not descriptive",
+    )
+    def test_function(a: int, b: int) -> int:
+        return a + b
 
-class WarningTests(TestCase):
-    """Tests of functions in warnings"""
+    expected = test_function(a=5, b=3)
+    got = test_function(x=5, y=3)
+    assert got == expected
 
-    def test_deprecate(self):
-        """test that deprecated decorator functions"""
 
-        def new_function():
-            return True
+def test_deprecated_args_emits_warning():
+    # Example target function to be decorated
+    @deprecated_args(
+        [("x", "a"), ("y", "b")],
+        version="a future release",
+        reason="x and y are not descriptive",
+    )
+    def test_function(a: int, b: int) -> int:
+        return a + b
 
-        @deprecate(new_function, "2023.3", "test deprecation")
-        def old_function():
-            return False
+    with pytest.deprecated_call():
+        test_function(x=5, y=3)
 
-        self.assertTrue(new_function())
-        self.assertTrue(old_function())
+    with pytest.deprecated_call():
+        test_function(a=5, y=3)
+    with pytest.deprecated_call():
+        test_function(x=5, b=3)
 
+    with pytest.warns(None) as record:  # verify no warnings for correct parameters
+        test_function(a=5, b=3)
+    assert not record
 
-if __name__ == "__main__":
-    main()
+
+def test_function_deprecated_args_maintains_docstring():
+    @deprecated_args(
+        [("x", "a"), ("y", "b")],
+        version="a future release",
+        reason="x and y are not descriptive",
+    )
+    def test_function(a: int, b: int) -> int:
+        """This is a test function"""
+        return a + b
+
+    assert test_function.__doc__ == "This is a test function"
+
+
+def test_method_deprecated_args():
+    class test_class:
```

Comment:
name issue as above for both the class and the method, just call the class `foo` and method `changed`